sympy/solvers/solveset.py

```diff
@@ -1117,3 +1120,525 @@ def linsolve(system, *symbols):
+    ordered tuple, i.e. the first & only argument to FiniteSet is a tuple of
+    solutions, which is ordered, & hence the returned solution is ordered.
+
+    Also note that solution could also have been returned as an ordered tuple,
+    FiniteSet is just a wrapper `{}` around the tuple. It has no other
+    significance except for the fact it is just used to maintain a consistent
+    output format throughout the solveset.
+
+    Examples
+    ========
+
+    >>> from sympy.core.symbol import symbols
+    >>> x, y = symbols('x, y', real = True)
+    >>> from sympy.solvers.solveset import substitution
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([]), [x, y])
+    {(-1, 1)}
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([x + 1]), [y, x])
+    EmptySet()
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([x - 1]), [y, x])
+    {(1, -1)}
+
+    """
+
+    from sympy.core.compatibility import ordered, default_sort_key
+    from sympy import Complement
+    from sympy.core.containers import Dict
+
+    def _unsolved_syms(eq, sort=False):
+        # get the unsolved symbol.
+        free = eq.free_symbols
+        unsolved = (free - set(known_symbols)) & set(all_symbols)
+        if sort:
+            unsolved = list(unsolved)
+            unsolved.sort(key=default_sort_key)
+        return unsolved
+
+    # sort such that equation with the fewest potential symbols is first.
+    # means eq with less variable first.
+    complements = {}
+    for eq in ordered(system, lambda _: len(_unsolved_syms(_))):
+        newresult = []
+        got_s = set()  # symbols solved in one iteration
+        hit = False
+        for r in result:
+            # update eq with everything that is known so far
+            eq2 = eq.subs(r)
+            unsolved_syms = _unsolved_syms(eq2, sort=True)
+            if not unsolved_syms:
+                if r:
+                    newresult.append(r)
+                break  # skip as it's independent of desired symbols
+            for s in unsolved_syms:
+                not_solvable = False
+                soln_imageset = None
+                try:
+                    soln = solveset_real(eq2, s)
+                    # If we want to add the complex solution also then
+                    # do the same with soln_complex. But it sometimes
+                    # we get complicated solution, so not adding it always
+                    # soln_complex = solveset_complex(eq2, s)
+                    if not soln:
+                        # If there is no real solution then check for
+                        # complex solution like old solve.
+                        soln = solveset_complex(eq2, s)
+                except NotImplementedError:
+                    continue
+                if isinstance(soln, Complement):
+                    # extract solution and complement
+                    complements[s] = soln.args[1]
+                    soln = soln.args[0]
```

Comment:
See my point 6 again. I don't understand the logic here. Please don't just assume and write anything, make sure you can justify, what you write.

I have already told you, don't write for particular case, It may not work for all cases!

```python
In [14]: soln = Complement(ImageSet(Lambda(x, x**2), N), Interval(0, oo))
In [15]: soln.args[0]
Out[15]: ImageSet(Lambda(x, x**2), <function N at 0x1045e6b70>)
```